rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs

```diff
@@ -0,0 +1,246 @@
+    > {
+        let pool = pool.consensus_pool.read().unwrap();
+        let nested = self.bouncer_factory.new_bouncer(&pool);
+
+        Box::new(move |id| nested(id.as_ref()))
+    }
+}
+
+#[derive(Clone)]
+pub struct FetchStrippedConsensusArtifact {
+    log: ReplicaLogger,
+    ingress_pool: ValidatedPoolReaderRef<SignedIngress>,
+    fetch_stripped: FetchArtifact<MaybeStrippedConsensusMessage>,
+    transport: Arc<dyn Transport>,
+    node_id: NodeId,
+    _metrics: FetchStrippedConsensusArtifactMetrics,
+}
+
+impl FetchStrippedConsensusArtifact {
+    pub fn new<Pool: ValidatedPoolReader<ConsensusMessage> + Send + Sync + 'static>(
+        log: ReplicaLogger,
+        rt: tokio::runtime::Handle,
+        consensus_pool: Arc<RwLock<Pool>>,
+        ingress_pool: ValidatedPoolReaderRef<SignedIngress>,
+        bouncer_factory: Arc<dyn BouncerFactory<ConsensusMessage, Pool>>,
+        metrics_registry: MetricsRegistry,
+        node_id: NodeId,
+    ) -> (impl Fn(Arc<dyn Transport>) -> Self, axum::Router) {
+        let ingress_pool_clone = ingress_pool.clone();
+        let consensus_pool_clone = consensus_pool.clone();
+
+        let router = super::download::build_axum_router(super::download::Pools {
+            consensus_pool: consensus_pool_clone,
+            ingress_pool: ingress_pool_clone,
+        });
+
+        let (fetch_stripped_fn, subrouter) = FetchArtifact::new(
+            log.clone(),
+            rt,
+            Arc::new(RwLock::new(ConsensusPoolWrapper { consensus_pool })),
+            Arc::new(BouncerFactoryWrapper { bouncer_factory }),
+            metrics_registry.clone(),
+        );
+
+        let router = axum::Router::new().merge(router).merge(subrouter);
+
+        let handler = move |transport: Arc<dyn Transport>| {
+            let fetch_stripped = fetch_stripped_fn(transport.clone());
+
+            Self {
+                log: log.clone(),
+                ingress_pool: ingress_pool.clone(),
+                fetch_stripped,
+                transport,
+                node_id,
+                _metrics: FetchStrippedConsensusArtifactMetrics::new(&metrics_registry),
+            }
+        };
+
+        (handler, router)
+    }
+}
+
+impl ArtifactAssembler<ConsensusMessage, MaybeStrippedConsensusMessage>
+    for FetchStrippedConsensusArtifact
+{
+    fn disassemble_message(&self, msg: ConsensusMessage) -> MaybeStrippedConsensusMessage {
+        msg.strip()
+    }
+
+    async fn assemble_message<P: Peers + Clone + Send + 'static>(
+        &self,
+        id: <MaybeStrippedConsensusMessage as IdentifiableArtifact>::Id,
+        artifact: Option<(MaybeStrippedConsensusMessage, NodeId)>,
+        peer_rx: P,
+    ) -> Result<(ConsensusMessage, NodeId), Aborted> {
+        // Download the Stripped message if it hasn't been pushed.
+        let (stripped_artifact, peer) = self
+            .fetch_stripped
+            .assemble_message(id.clone(), artifact, peer_rx.clone())
```

Comment:
Maybe I'm confused, why does the function `assemble_message` return a `stripped_artifact`?

Reply:
Since stripped artifacts may still get advertised it can happen that you only receive the `StrippedMessageId`. So the first step is to download the `StrippedMessage` by doing an rpc. But yes, the naming could be better.

Reply:
I see, thanks!